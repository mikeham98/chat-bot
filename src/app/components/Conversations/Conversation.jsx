import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from '../../../themes/components/conversations.scss';

export default class Conversation extends React.PureComponent {
    returnClassNames() {
        const {selected, unread} = this.props;
        const classNames = [styles.conversationWrapper];
        if (selected) {
            classNames.push(styles.conversationSelected);
        }
        if (unread) {
            classNames.push(styles.conversationUnread);
        }
        return classnames(classNames);
    }

    render() {
        const {onClick, image, name, previewMessage} = this.props;
        return (
            <div onClick={onClick} className={this.returnClassNames()}>
                <img src={image}/>
                <div className={styles.conversationPreview}>
                    <span className={styles.conversationPreviewName}>{name}</span>
                    {previewMessage &&
                    <span className={styles.conversationPreviewMessage}>{previewMessage}</span>}
                </div>
            </div>
        );
    }
};

Conversation.defaultProps = {
    previewMessage: ''
};

Conversation.propTypes = {
    selected: PropTypes.bool,
    unread: PropTypes.bool,
    onClick: PropTypes.func,
    previewMessage: PropTypes.string,
    image: PropTypes.string,
    name: PropTypes.string,
};