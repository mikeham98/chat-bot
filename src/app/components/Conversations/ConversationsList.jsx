import React from 'react';
import PropTypes from 'prop-types';
import Conversation from "./Conversation";
import isPopulatedArray from "../../util/isPopulatedArray";
import styles from "../../../themes/components/conversations.scss";
import {ConversationsPropTypes} from "../../config/propTypes";

export default class ConversationsList extends React.PureComponent {
    returnConversations() {
        const {conversations} = this.props;
        if (isPopulatedArray(conversations)) {
            return conversations.map(conversation => {
                return (
                    <Conversation
                        key={conversation.id}
                        unread={typeof conversation.read === 'boolean' && !conversation.read}
                        previewMessage={conversation.latest ? conversation.latest.message : ''}
                        selected={conversation.selected}
                        name={conversation.profile.name}
                        image={conversation.profile.img}
                        onClick={() => this.props.onClickConversation(conversation.id)}
                    />
                );
            });
        }
    }

    render() {
        return (
            <div className={styles.conversationListWrapper}>
                {this.returnConversations()}
            </div>
        )
    }
}

ConversationsList.propTypes = {
    conversations: ConversationsPropTypes,
    onClickConversation: PropTypes.func.isRequired,
};