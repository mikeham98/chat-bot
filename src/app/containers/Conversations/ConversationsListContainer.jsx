import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import ConversationsList from '../../components/Conversations/ConversationsList';
import {getConversationList, setCurrentConversation} from '../../actions/conversations/index.actions';
import {conversationSelector} from '../../selectors/conversationSelector';
import {ConversationsPropTypes} from "../../config/propTypes";

export class ConversationsListContainer extends React.PureComponent {
    constructor(props) {
        super(props);
        this.onClickConversation = this.onClickConversation.bind(this);
    }

    componentDidMount() {
        this.props.getConversationList();
        // update backend to so that read is true on the initial load
        // (because the conversation selected by default is the first one)
        this.props.setCurrentConversation('1');
    }

    onClickConversation(conversationId) {
        const {id} = this.props.currentConversation;
        if (id !== conversationId) {
            this.props.setCurrentConversation(conversationId);
        }
    }

    render() {
        return (
            <ConversationsList
                conversations={this.props.conversations}
                onClickConversation={this.onClickConversation}
            />
        )
    }
}

ConversationsListContainer.propTypes = {
    conversations: ConversationsPropTypes,
    getConversationList: PropTypes.func.isRequired,
    setCurrentConversation: PropTypes.func.isRequired,
};

/* istanbul ignore next */
const mapStateToProps = (state) => {
    return conversationSelector(state);
};

export default connect(mapStateToProps, {getConversationList, setCurrentConversation})(ConversationsListContainer)