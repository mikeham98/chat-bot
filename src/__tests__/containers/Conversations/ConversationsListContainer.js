import React from 'react';
import {shallow} from 'enzyme';
import {ConversationsListContainer} from '../../../app/containers/Conversations/ConversationsListContainer';
import ConversationsList from '../../../app/components/Conversations/ConversationsList';

describe('ConversationsListContainer', () => {

    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if (!shallowedWrapper) {
            shallowedWrapper = shallow(<ConversationsListContainer {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            getConversationList: jest.fn(),
            setCurrentConversation: jest.fn(),
            currentConversationId: 3,
            conversations: [1, 2, 3]
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return ConversationsList component', () => {
            expect(wrapper()).toMatchSnapshot();
        });
    });

    describe('functionality', () => {
        describe('props', () => {
            describe('getConversationList', () => {
                it('should be called on mount', () => {
                    expect(props.getConversationList).not.toHaveBeenCalled();
                    wrapper();
                    expect(props.getConversationList).toHaveBeenCalledTimes(1);
                });
            });
            describe('setCurrentConversation', () => {
                it('should be called on call of prop onClickConversation on ConversationList', () => {
                    expect(props.setCurrentConversation).not.toHaveBeenCalled();
                    wrapper().find(ConversationsList).first().props().onClickConversation(2);
                    expect(props.setCurrentConversation).toHaveBeenCalledWith(2);
                });
                it('should not be called on call of prop onClickConversation on ConversationList ' +
                    'due to matching conversationId', () => {
                    expect(props.setCurrentConversation).not.toHaveBeenCalled();
                    wrapper().find(ConversationsList).first().props().onClickConversation(3);
                    expect(props.setCurrentConversation).not.toHaveBeenCalled();
                });
            });
        });
    });
});