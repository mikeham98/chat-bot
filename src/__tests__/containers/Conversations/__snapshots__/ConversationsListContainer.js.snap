// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationsListContainer render should return ConversationsList component 1`] = `
<ConversationsList
  conversations={
    Array [
      1,
      2,
      3,
    ]
  }
  onClickConversation={[Function]}
/>
`;
